src/workerd/api/sockets.c++

```diff
@@ -128,4 +134,38 @@ void Socket::handleProxyStatus(
   result.markAsHandled();
 }
 
+void Socket::handleReadableEof(jsg::Lock& js) {
+  // Listen for EOF on the ReadableStream.
+  KJ_ASSERT_NONNULL(readable->eofResolverPair).promise.then(js,
+      JSG_VISITABLE_LAMBDA((ref=JSG_THIS), (ref), (jsg::Lock& js) {
+    return ref->maybeCloseWriteSide(js);
+  })).markAsHandled();
+}
+
+jsg::Promise<void> Socket::maybeCloseWriteSide(jsg::Lock& js) {
+  // When `allowHalfOpen` is set to true then we do not automatically close the write side on EOF.
+  // The default value for `allowHalfOpen` is also false.
+  KJ_IF_MAYBE(opts, options) {
+    if (opts->allowHalfOpen) {
+      return js.resolvedPromise();
+    }
+  }
+
+  // We want to close the socket, but only after its WritableStream has been flushed. We do this
+  // below by calling `close` on the WritableStream which ensures that any data pending on it
+  // is flushed. Then once the `close` either completes or fails we can be sure that any data has
+  // been flushed.
+  return writable->getController().close(js).catch_(js,
+      [](jsg::Lock& js, workerd::jsg::V8Ref<v8::Value> exc) {
+    // A failure to close the WritableStream can indicate one of these things:
+    //   * The WritableStream hasn't been attached.
+    //   * The WritableStream has already been released or closed.
+    //
+    // We only want to ensure that the writable stream is flushed before closing the socket.
+    // With the above we are certain that it was flushed, so we are safe to close.
+  }).then(js, JSG_VISITABLE_LAMBDA((ref=JSG_THIS), (ref), (jsg::Lock& js) {
+    ref->closeFulfiller->resolver.resolve();
```

Comment:
This delivers the close event only after `close()` has completed, but is that really what we want? I could see an argument that we should really deliver the close event immediately when we call `close()`, so that the application can know "this is closed now, I should not keep writing to it". Otherwise the app might attempt writes in between which will fail.

That said, since this is resolving a promise, there is still a race condition either way, if the app is actively writing at the time that the close is delivered.

(More reasons why I find `allowHalfOpen = false` to be really weird.)